pkg/aws/eni/node.go

```diff
@@ -538,3 +539,49 @@ func (n *Node) GetMaximumAllocatableIPv4() int {
 	// Return the maximum amount of IP addresses allocatable on the instance
 	return (limits.Adapters - firstInterfaceIndex) * limits.IPv4
 }
+
+// GetMinimumAllocatableIPv4 returns the minimum amount of IPv4 addresses that
+// must be allocated to the instance.
+func (n *Node) GetMinimumAllocatableIPv4() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
```

Comment:
My knowledge of mutexes in Golang is limited, so this is probably a naive question. Isn't the `n.mutex.RLock()` above defeated by the `n.mutex.Lock()` that you take before calling the `GetMinimumAllocatableIPv4()` function?

Reply:
Good question. Looking at the code, I see why you're confused. The answer doesn't have to do with anything specific to Golang mutexes. The `Node` in `GetMinimumAllocatableIPv4()` refers to the `pkg/aws/eni.Node` type, whereas at the site of the caller where the `Lock()` takes place as you observed is of type `pkg/ipam.Node`. These mutexes are locking different structures, although they are named the same, so it can be confusing to misleading to readers.  :)

Reply:
Ah I see, ok thank you! :)